### PR TITLE
MBS-9079 Helper for namespaced resources in application

### PR DESCRIPTION
**Pull request by eugene-krivobokov**
Added plugin to replace unambiguous non-namespaced R classes to namespaced ones (android.namespacedRClass).
See cases in tests.

### CODE REVIEW

File: docs/content/docs/checks/AndroidLint.md [hunk 1]
```diff
@@ -45,3 +45,5 @@ How to start:
 
 - [Static Analysis with Android Lint by Tor Norbye (mDevCamp 2019)](https://slideslive.com/38916502) 
 - [Sample project](https://github.com/googlesamples/android-custom-lint-rules)
+- Install [PsiViewer IDE plugin](https://plugins.jetbrains.com/plugin/227-psiviewer/)
```
Comment (sergioboi) [inline docs/content/docs/checks/AndroidLint.md:48]: Why did you add these links to block about Lint?

File: subprojects/gradle/namespaced-resources-fixer/src/test/kotlin/com/avito/android/plugin/FileResourcesFixerTest.kt [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+            class Screen {
+                val id = R.id.container
+            }
+        """.trimIndent(),
+            app, libraries
+        )
+        assertThat(result).isEqualTo(
+            """
+            import com.avito.android.R
```
Comment (sergioboi) [inline subprojects/gradle/namespaced-resources-fixer/src/test/kotlin/com/avito/android/plugin/FileResourcesFixerTest.kt]: Why do we need that import in the result?

File: subprojects/gradle/namespaced-resources-fixer/src/test/kotlin/com/avito/android/plugin/FileResourcesFixerTest.kt [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+            class Element() : ViewElement {
+                val icon = element<ViewElement>(ViewMatchers.withId(R.id.icon))
+            }
+        """.trimIndent(),
+            app, libraries
+        )
+        assertThat(result).isEqualTo(
+            """
+            import androidx.test.espresso.matcher.ViewMatchers
+            import com.avito.android.R
```
Comment (sergioboi) [inline subprojects/gradle/namespaced-resources-fixer/src/test/kotlin/com/avito/android/plugin/FileResourcesFixerTest.kt]: The same ^

File: subprojects/gradle/namespaced-resources-fixer/src/test/kotlin/com/avito/android/plugin/FileResourcesFixerTest.kt [hunk 4]
```diff
@@ -0,0 +1,147 @@
+package com.avito.android.plugin
+
+import com.google.common.truth.Truth.assertThat
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.io.TempDir
+import java.io.File
+
+class FileResourcesFixerTest {
```
Comment (sergioboi) [inline subprojects/gradle/namespaced-resources-fixer/src/test/kotlin/com/avito/android/plugin/FileResourcesFixerTest.kt:9]: I expect to see how Fixer will work with aliases in imports e.g.
`com.avito.android.R as app_R`

File: subprojects/gradle/namespaced-resources-fixer/src/main/kotlin/com/avito/android/plugin/FileResourcesFixer.kt [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+ */
+internal class FileResourcesFixer {
+
+    private val compiler = KtCompiler()
+
+    /**
+     * @return new file content or null if nothing changed
+     */
+    fun fixMergedResources(file: File, app: AndroidModule, libraries: List<AndroidModule>): String? {
+        val ktFile = compiler.compile(file.toPath(), file.toPath())
```
Comment (sergioboi) [inline subprojects/gradle/namespaced-resources-fixer/src/main/kotlin/com/avito/android/plugin/FileResourcesFixer.kt]: Do you mean a classFile? Or compiler is a parser from text to kotlinPsi?

File: subprojects/gradle/namespaced-resources-fixer/src/main/kotlin/com/avito/android/plugin/FileResourcesFixer.kt [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+            }
+        }
+        return replacedModules
+    }
+
+    private class MergedRClassReferencesFinder(
+        private val classImport: KtImportDirective
+    ) : KtTreeVisitorVoid() {
+
+        var references = mutableListOf<PsiElement>()
```
Comment (sboishtyan) [inline subprojects/gradle/namespaced-resources-fixer/src/main/kotlin/com/avito/android/plugin/FileResourcesFixer.kt:65]: Consider that local val returned from visitElement

File: subprojects/gradle/namespaced-resources-fixer/src/main/kotlin/com/avito/android/plugin/FileResourcesFixer.kt [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+
+        override fun visitElement(element: PsiElement) {
```
Comment (sboishtyan) [inline subprojects/gradle/namespaced-resources-fixer/src/main/kotlin/com/avito/android/plugin/FileResourcesFixer.kt:67]: `visitElement` looks common
I prefer see concrete name `findReferences` or something similar

File: subprojects/gradle/namespaced-resources-fixer/src/main/kotlin/com/avito/android/plugin/FileResourcesFixer.kt [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+            if (element.isReferenceTo(classImport)) {
+                references.add(element)
+            }
+            super.visitElement(element)
+        }
+
+        private fun PsiElement.isReferenceTo(importElement: KtImportDirective): Boolean {
```
Comment (sergioboi) [inline subprojects/gradle/namespaced-resources-fixer/src/main/kotlin/com/avito/android/plugin/FileResourcesFixer.kt:74]: Add some Psi documentation reference please
I didn't understand
- what isDotReferenceExpression means
- what partialDotReferenceExpression means
- what insideImport means

File: subprojects/gradle/namespaced-resources-fixer/src/main/kotlin/com/avito/android/plugin/FileResourcesFixer.kt [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+    }
+
+    private fun findMergedRImport(file: KtFile, app: AndroidModule): KtImportDirective? {
+        val imports: List<KtImportDirective> = file.importList?.imports?.toList() ?: return null
+        return imports.firstOrNull {
+            it.importPath.toString().startsWith(app.packageId)
+        }
+    }
+
+    private fun findUniqueResources(modules: List<AndroidModule>): List<ResourceId> {
```
Comment (sergioboi) [inline subprojects/gradle/namespaced-resources-fixer/src/main/kotlin/com/avito/android/plugin/FileResourcesFixer.kt:101]: Is it class from detekt?

File: subprojects/gradle/namespaced-resources-fixer/src/main/kotlin/com/avito/android/plugin/FileResourcesFixer.kt [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+    }
+
+    private fun replaceReferences(
+        rClassReferences: List<PsiElement>,
+        uniqueResources: List<ResourceId>
+    ): Set<AndroidModule> {
+        val replacedModules = mutableSetOf<AndroidModule>()
+        rClassReferences.forEach { refElement ->
+            val resource = uniqueResources.find { resource ->
+                refElement.text.endsWith(".id.${resource.id}")
```
Comment (sergioboi) [inline subprojects/gradle/namespaced-resources-fixer/src/main/kotlin/com/avito/android/plugin/FileResourcesFixer.kt:50]: It will work incorrect with identical ids from different modules
```
import com.android.app.R as app_R
import com.android.library.R as library_R

val r = app_R.id.one
val r1 = library_R.id.one
```

File: subprojects/gradle/namespaced-resources-fixer/src/main/kotlin/com/avito/android/plugin/FileResourcesFixer.kt [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+            """.trimIndent()
+        )
+        return file.children.find { it is PsiWhiteSpace }!! as PsiWhiteSpace
+    }
+
+    private fun createElementForResourceRef(resource: ResourceId): PsiElement {
+        val alias = importAliasForResource(resource.module)
+
+        val file = compiler.createKtFile("val id = $alias.id.${resource.id}")
+        return file.children.find { it is KtProperty }!!.children[0]
```
Comment (sergioboi) [inline subprojects/gradle/namespaced-resources-fixer/src/main/kotlin/com/avito/android/plugin/FileResourcesFixer.kt:166]: Sad that we have to create a whole file.
Do you dig about that?

File: docs/content/docs/checks/AndroidLint.md (hunk 1 above)
Comment (eugene-krivobokov) [reply docs/content/docs/checks/AndroidLint.md:48]: - They contain useful information for writing custom lint checks.
- We don't have IDE plugins yet. They will have common things

File: subprojects/gradle/namespaced-resources-fixer/src/main/kotlin/com/avito/android/plugin/FileResourcesFixer.kt [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+import org.jetbrains.kotlin.psi.KtImportDirective
+import org.jetbrains.kotlin.psi.KtImportList
+import org.jetbrains.kotlin.psi.KtProperty
+import org.jetbrains.kotlin.psi.KtTreeVisitorVoid
+import java.io.File
+
+/**
+ * Replace unambiguous R class references to Application by libraries R classes
+ */
+internal class FileResourcesFixer {
```
Comment (eugene-krivobokov) [inline subprojects/gradle/namespaced-resources-fixer/src/main/kotlin/com/avito/android/plugin/FileResourcesFixer.kt:17]: The main idea:
- Find unique resources in libraries
- Find references to a merged R class file
- Replace them by references to libraries' R classes
- Add missing imports

File: subprojects/gradle/namespaced-resources-fixer/src/main/kotlin/com/avito/android/plugin/FileResourcesFixer.kt (hunk 5 above)
Comment (eugene-krivobokov) [reply subprojects/gradle/namespaced-resources-fixer/src/main/kotlin/com/avito/android/plugin/FileResourcesFixer.kt]: It's a `org.jetbrains.kotlin.psi.KtFile`.
As far as I understand, it's a lightweight AST, not compiled.

Comment (eugene-krivobokov) [reply subprojects/gradle/namespaced-resources-fixer/src/main/kotlin/com/avito/android/plugin/FileResourcesFixer.kt]: moved to `parse` extension function to reflect this intention

File: subprojects/gradle/namespaced-resources-fixer/src/test/kotlin/com/avito/android/plugin/FileResourcesFixerTest.kt (hunk 4 above)
Comment (eugene-krivobokov) [reply subprojects/gradle/namespaced-resources-fixer/src/test/kotlin/com/avito/android/plugin/FileResourcesFixerTest.kt:9]: Added new test `replace library id - simple property with alias to R`

File: subprojects/gradle/namespaced-resources-fixer/src/main/kotlin/com/avito/android/plugin/FileResourcesFixer.kt (hunk 10 above)
Comment (eugene-krivobokov) [reply subprojects/gradle/namespaced-resources-fixer/src/main/kotlin/com/avito/android/plugin/FileResourcesFixer.kt:50]: Added new test `replace library id - two simple properties with the same name`

File: subprojects/gradle/namespaced-resources-fixer/src/test/kotlin/com/avito/android/plugin/FileResourcesFixerTest.kt (hunk 2 above)
Comment (eugene-krivobokov) [reply subprojects/gradle/namespaced-resources-fixer/src/test/kotlin/com/avito/android/plugin/FileResourcesFixerTest.kt]: I want to postpone this optimization with a merged R class.

It's a rare case when we have only unambiguous usages and can remove merged R class after such optimization.

As a next improvement, I want to rename it:

```kotlin
import com.avito.android.R as deprecated_merged_R
```

File: subprojects/gradle/namespaced-resources-fixer/src/test/kotlin/com/avito/android/plugin/FileResourcesFixerTest.kt (hunk 3 above)
Comment (eugene-krivobokov) [reply subprojects/gradle/namespaced-resources-fixer/src/test/kotlin/com/avito/android/plugin/FileResourcesFixerTest.kt]: https://github.com/avito-tech/avito-android/pull/510#discussion_r459052243
Answered there

File: subprojects/gradle/namespaced-resources-fixer/src/main/kotlin/com/avito/android/plugin/FileResourcesFixer.kt (hunk 9 above)
Comment (eugene-krivobokov) [reply subprojects/gradle/namespaced-resources-fixer/src/main/kotlin/com/avito/android/plugin/FileResourcesFixer.kt:101]: No, they are just DTOs in this plugin:

```kotlin
internal data class AndroidModule(val packageId: String, val ids: List<String>)

internal data class ResourceId(val id: String, val module: AndroidModule)
```

File: subprojects/gradle/namespaced-resources-fixer/src/main/kotlin/com/avito/android/plugin/FileResourcesFixer.kt (hunk 11 above)
Comment (eugene-krivobokov) [reply subprojects/gradle/namespaced-resources-fixer/src/main/kotlin/com/avito/android/plugin/FileResourcesFixer.kt:166]: Have tried PsiElementFactory and PsiJavaParserFacade with no luck.
Maybe miss something but don't want to spend more time for this optimization.

File: subprojects/gradle/namespaced-resources-fixer/src/main/kotlin/com/avito/android/plugin/FileResourcesFixer.kt (hunk 8 above)
Comment (eugene-krivobokov) [reply subprojects/gradle/namespaced-resources-fixer/src/main/kotlin/com/avito/android/plugin/FileResourcesFixer.kt:74]: import

PsiViewer is a good way to check a state.
Psi is a tree with different types of nodes.

These local variables are just named checks against psi elements, nothing more.

Comment (eugene-krivobokov) [reply subprojects/gradle/namespaced-resources-fixer/src/main/kotlin/com/avito/android/plugin/FileResourcesFixer.kt:74]: property

File: subprojects/gradle/namespaced-resources-fixer/src/main/kotlin/com/avito/android/plugin/FileResourcesFixer.kt (hunk 9 above)
Comment (eugene-krivobokov) [reply subprojects/gradle/namespaced-resources-fixer/src/main/kotlin/com/avito/android/plugin/FileResourcesFixer.kt:101]: In short, these models describe Gradle projects.

File: subprojects/gradle/namespaced-resources-fixer/src/main/kotlin/com/avito/android/plugin/FileResourcesFixer.kt (hunk 11 above)
Comment (eugene-krivobokov) [reply subprojects/gradle/namespaced-resources-fixer/src/main/kotlin/com/avito/android/plugin/FileResourcesFixer.kt:166]: Here is a starting point: 
https://jetbrains.org/intellij/sdk/docs/basics/architectural_overview/modifying_psi.html?search=psiWHitespace#creating-the-new-psi

**Review by dsvoronin — APPROVED** [review]

File: subprojects/gradle/namespaced-resources-fixer/src/main/kotlin/com/avito/android/plugin/FileResourcesFixer.kt [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+
+    private val compiler = KtCompiler()
+
+    /**
+     * @return new file content or null if nothing changed
+     */
+    fun fixMergedResources(file: File, app: AndroidModule, libraries: List<AndroidModule>): String? {
+        val ktFile = compiler.parse(file)
```
Comment (int02h) [inline subprojects/gradle/namespaced-resources-fixer/src/main/kotlin/com/avito/android/plugin/FileResourcesFixer.kt:25]: Did you measure how long does it take to parse and fix all files in the project?

**Review by int02h — APPROVED** [review]

**Review by int02h — APPROVED** [review]

File: subprojects/gradle/namespaced-resources-fixer/src/main/kotlin/com/avito/android/plugin/FileResourcesFixer.kt (hunk 13 above)
Comment (eugene-krivobokov) [reply subprojects/gradle/namespaced-resources-fixer/src/main/kotlin/com/avito/android/plugin/FileResourcesFixer.kt:25]: About 40 seconds for the whole test source sets in the application module. 
The most expensive part is to build all modules.

File: subprojects/gradle/namespaced-resources-fixer/src/main/kotlin/com/avito/android/plugin/FileResourcesFixer.kt (hunk 11 above)
Comment (eugene-krivobokov) [reply subprojects/gradle/namespaced-resources-fixer/src/main/kotlin/com/avito/android/plugin/FileResourcesFixer.kt:166]: Fixed here: https://github.com/avito-tech/avito-android/pull/514/files
